dbt/linker.py

```diff
@@ -16,6 +18,104 @@ def from_file(graph_file):
+        # and worker thread reservations
+        self.in_progress = set()
+        # things that are in the queue
+        self.queued = set()
+        # this lock controls most things
+        self.lock = threading.Lock()
+        # store the number of descendants for each node.
+        self._descendants = self._calculate_descendants()
+        # populate the initial queue
+        self._find_new_additions()
+
+    def get_node(self, node_id):
+        return self.manifest.nodes[node_id]
+
+    def _include_in_cost(self, node_id):
+        node = self.get_node(node_id)
+        if not dbt.utils.is_blocking_dependency(node):
+            return False
+        if node.get_materialization() == 'ephemeral':
+            return False
+        return True
+
+    def _calculate_descendants(self):
+        """We could do this in one pass over the graph instead of
+        len(self.graph) passes but this is easy. For large graphs this may hurt
+        performance.
+        """
+        costs = {}
+        for node in self.graph.nodes():
+            cost = len([
+                d for d in nx.descendants(self.graph, node)
+                if self._include_in_cost(d)
+            ])
+            costs[node] = cost
+        return costs
+
+    def get(self, block=True, timeout=None):
+        """Get a node off the inner priority queue. By default, this blocks.
+        """
+        # It's unsafe to call get() and empty() from different threads, so
+        # don't do that
+        _, node_id = self.inner.get(block=block, timeout=timeout)
+        with self.lock:
+            self._mark_in_progress(node_id)
+        return self.get_node(node_id)
+
+    def __len__(self):
+        with self.lock:
+            return len(self.graph) - len(self.in_progress)
+
+    def empty(self):
+        """The graph queue is 'empty' if it all remaining nodes are in progress
+        """
+        return len(self) == 0
+
+    def _already_known(self, node):
+        """Callers must hold the lock
+        """
+        return node in self.in_progress or node in self.queued
+
+    def _find_new_additions(self):
+        """Find any nodes in the graph that need to be added to the internal
+        queue.
+
+        Note: callers must hold the lock
+        """
+        for node, in_degree in self.graph.in_degree_iter():
+            if not self._already_known(node) and in_degree == 0:
+                # lower score = more descendants = higher priority
+                score = -1 * self._descendants[node]
+                self.inner.put((score, node))
+                self.queued.add(node)
+
+    def mark_done(self, task_id):
+        with self.lock:
+            self.in_progress.remove(task_id)
+            self.graph.remove_node(task_id)
+            self._find_new_additions()
+            self.inner.task_done()
+
+    def _mark_in_progress(self, task_id):
+        self.queued.remove(task_id)
+        self.in_progress.add(task_id)
```

Comment:
I don't really feel like the lock management is that hard here to require an `RLock`, but maybe I'm overly reluctant to pull those in. The biggest issue is making sure not to lock around `self.inner.get` and triggering deadlocks, which `RLock` won't really help with. I will add a docstring to this method indicating it needs to hold the lock though, as that is important.

I'm happy to discuss my thoughts on `RLock` as they're complicated and informed by my many threading-related mistakes in life, but in short: it's a dangerous tool that can easily make you complacent about locking, leading to mystery deadlocks. With only 3 places that take the lock (two of which have to be the same thread!), a lack of forward progress is much easier to debug. When they're at all easy to avoid, I prefer to do so.

Reply:
Also, I'm not sure we want to document it, but all the underscore methods in this class require that you hold the lock (the only ones that don't are called from `__init__` where the initializing thread has exclusive access to begin with). Once you're inside an `_*` method, you can assume the lock is held and behave accordingly.